Add IsAccessible(method) that provides a message when not accessible

blairconrad opened the pull request:
Fixes #235.

I may not have put the code in the best places.
Note that only the "references Castle.Core/strong-named" path is tested in the unit tests. If you'd like, I can add tests for the other paths, but believe they would require additional test projects.
I have performed manual testing for the other paths.

Thanks for considering this PR. I'm happy to make changes you might like.

blairconrad commented:
Hi, @jonorossi. I hate to be a bother, but I was wondering if you had any comments or concerns.

jonorossi commented on src/Castle.Core/DynamicProxy/Internal/InternalsUtil.cs:
```diff
@@ -96,5 +98,52 @@ public static bool IsAccessible(this MethodBase method)
+#else
+// .NET Core does not provide an API to do this, so we just fall back to the solution that will definitely work.
+// After all it is just an exception message.
+      return false;
```
The indentation here is broken, some lines with spaces rather than tabs and the comment has lost its indentation.

blairconrad replied:
:astonished: How embarrassing. I did not notice. Also, I'm surprised editorconfig did not save me from this. But still, ultimately my fault!

jonorossi commented on src/Castle.Core/DynamicProxy/ProxyUtil.cs:
```diff
@@ -80,5 +82,36 @@ public static bool IsProxyType(Type type)
+		/// <param name="method">The method to check</param>
+		/// <param name="message">If the method is accessible, <c>null</c>; otherwise, an explanation of why the method is not accessible</param>
+		/// <returns><c>true</c> if the method is accessible, <c>false</c> otherwise</returns>
+		public static bool IsAccessible(MethodBase method, out string message)
```
While we are here I'm thinking it would be worthwhile adding a `public static bool IsAccessible(MethodBase method)` (non-extension method) overload that just calls the InternalsUtil extension method.

blairconrad replied:
Done!

blairconrad commented on src/Castle.Core/DynamicProxy/Internal/InternalsUtil.cs:
```diff
@@ -96,5 +98,52 @@ public static bool IsAccessible(this MethodBase method)
+			// After all it is just an exception message.
+			return false;
+#endif
+		}
```
I think the spacing's better now. Not sure what I did before. Thanks/sorry.

jonorossi commented:
@blairconrad looks great, thanks for making the changes. Just remember to update the changelog before squishing.

blairconrad commented:
> Just reviewed the changes and added a couple of comments, **could you also update the changelog**. We'll get this released as soon as we get the build refactor done.

I didn't even see this comment yesterday. So sorry about the changelog.

Updated. Squashed everything. You can compare with https://github.com/blairconrad/Core/tree/internal-method-message-unsquashed, if you've a mind to. It has the changes x`that you approved.

blairconrad commented:
Thanks, @jonorossi!

jonorossi commented:
Unfortunately the Mono build failed, it seems mcs is pickier with its warnings:

http://builds.castleproject.org/viewLog.html?buildId=14563&buildTypeId=Core_MasterMono&tab=buildLog

```
DynamicProxy.Tests/ProxyUtilTestCase.cs(82,36): error CS0414: Warning as Error: The private field `Castle.DynamicProxy.Tests.ProxyUtilTestCase.AccessibleMethods' is assigned but its value is never used
DynamicProxy.Tests/ProxyUtilTestCase.cs(90,36): error CS0414: Warning as Error: The private field `Castle.DynamicProxy.Tests.ProxyUtilTestCase.InaccessibleMethods' is assigned but its value is never used
```

blairconrad commented:
How embarrassing, @jonorossi. I took a shortcut by not running the MONO build. I can certainly address. But not for hours. Unless you don't mind me applying an obvious fix and then hoping it works out.

jonorossi commented:
Seriously it's not a big deal, I wouldn't call it a shortcut, Mono just gets annoying. Will be good once we've got the CI builds on pull requests.

I don't mind if you want to go for it now, or take a proper look later.

jonorossi commented:
Looks like these two tests are failing in the .NET Core build too:
- ClassProxyConstructorsTestCase.Cannot_proxy_inaccessible_class
- ProxyUtilTestCase.IsAccessibleWithReason_Inaccessible_Method_Populates_ReasonMethodIsNotAccessible("APrivateMethod")

blairconrad commented:
Oh, man. You're not going to ask me back!
Investigating.

jonorossi commented:
LOL, you are always welcome. This really just highlights how much our automated processes suck right now.

blairconrad commented:
The .NET Core failures suggest that the .NET Core build of Castle.Core.Tests isn't strong-named, or that the code can't tell that they are strong-named. Even though I see the snk in the project.json. And `sn -T` tells me it's signed. I'm gonna need some time with that.
